backend/v3/storage/database/repository/project_grant.go

```diff
@@ -0,0 +1,300 @@
+
+type projectGrant struct{}
+
+// ProjectGrantRepository manages project grants.
+func ProjectGrantRepository() domain.ProjectGrantRepository {
+	return new(projectGrant)
+}
+
+func (p projectGrant) Get(ctx context.Context, client database.QueryExecutor, opts ...database.QueryOption) (*domain.ProjectGrant, error) {
+	opts = append(opts,
+		database.WithGroupBy(p.InstanceIDColumn(), p.IDColumn()),
+	)
+
+	builder, err := p.prepareQuery(opts)
+	if err != nil {
+		return nil, err
+	}
+	return getOne[domain.ProjectGrant](ctx, client, builder)
+}
+
+func (p projectGrant) List(ctx context.Context, client database.QueryExecutor, opts ...database.QueryOption) ([]*domain.ProjectGrant, error) {
+	opts = append(opts,
+		database.WithGroupBy(p.InstanceIDColumn(), p.IDColumn()),
+	)
+
+	builder, err := p.prepareQuery(opts)
+	if err != nil {
+		return nil, err
+	}
+	return getMany[domain.ProjectGrant](ctx, client, builder)
+}
+
+const insertProjectGrantStmt = `INSERT INTO zitadel.project_grants(
+	instance_id, id, project_id, granting_organization_id, granted_organization_id, state
+)
+VALUES ($1, $2, $3, $4, $5, $6)
+RETURNING created_at, updated_at`
+
+const insertProjectGrantWithRolesStmt = `WITH added_roles AS (
+	INSERT INTO zitadel.project_grant_roles (
+		instance_id, grant_id, project_id, key
+	)
+	VALUES ($1, $2, $3, unnest($7::text[]))
+) ` + insertProjectGrantStmt
+
+func (p projectGrant) Create(ctx context.Context, client database.QueryExecutor, projectGrant *domain.ProjectGrant) error {
+	var builder *database.StatementBuilder
+	if len(projectGrant.RoleKeys) == 0 {
+		builder = database.NewStatementBuilder(insertProjectGrantStmt,
+			projectGrant.InstanceID,
+			projectGrant.ID,
+			projectGrant.ProjectID,
+			projectGrant.GrantingOrganizationID,
+			projectGrant.GrantedOrganizationID,
+			projectGrant.State,
+		)
+	} else {
+		builder = database.NewStatementBuilder(insertProjectGrantWithRolesStmt,
+			projectGrant.InstanceID,
+			projectGrant.ID,
+			projectGrant.ProjectID,
+			projectGrant.GrantingOrganizationID,
+			projectGrant.GrantedOrganizationID,
+			projectGrant.State,
+			projectGrant.RoleKeys,
+		)
+	}
+
+	if err := client.QueryRow(ctx, builder.String(), builder.Args()...).
+		Scan(&projectGrant.CreatedAt, &projectGrant.UpdatedAt); err != nil {
+		return err
+	}
+	return nil
+}
+
+const queryUpdateProjectGrantRoleStmt = `SELECT instance_id, id, project_id, unnest($1::text[]) as key from zitadel.project_grants`
+
+const updateProjectGrantRoleStmt = `removed_roles AS (
+    DELETE FROM zitadel.project_grant_roles as pgr
+    WHERE NOT EXISTS (
+		SELECT 1
+			FROM pg
+			WHERE pg.instance_id = pgr.instance_id
+			AND pg.id = pgr.grant_id
+			AND pg.key = pgr.key 
+	)
+), added_roles AS (
+	INSERT INTO zitadel.project_grant_roles (
+		instance_id, grant_id, project_id, key
+	)
+	SELECT instance_id, id, project_id, key FROM pg
+	ON CONFLICT (instance_id, grant_id, key) DO NOTHING
+)
+UPDATE zitadel.project_grants 
+SET updated_at = now()
```

Comment:
This would be wrong if the roles are set through events, we need the event.CreatedAt() then